### DISCUSSION TITLE
Some questions regarding VGMTrans (documentation related)

### DISCUSSION THREAD

**Discussion by Fulminis-ictus** [vgmtrans/vgmtrans#219] [General]
I'm currently researching the PS1's sound system for a paper and loveemu let me know that this would be a good place to ask these things. I hope you can help me out!

I was wondering if there's any documentation with which a coding knowledge lacking layman like myself can figure out how VGMtrans interprets the bytes of an SEQ file to display what command they correlate to (in the same way how for example MIDI interprets „[9, channel] [note] [velocity]“ as a Note-On command). I've checked out SeqEvent.cpp but didn't quite understand how all of it translated to hex code.

In the same vein, is there any documentation on how some custom PS1 audio drivers work (what drivers there are, which commands they accept, what those commands do etc.) or what the differences to generic SEQ are? I couldn't find any info whatsoever on them except for loveemu's blogpost on Dragon Quest's SEQq format (http://loveemu.hatenablog.com/entry/20140103/PlayStation_Dragon_Quest_Format).

The only two sources I could find on SEQ are unfortunately only loveemu's blog post (http://loveemu.hatenablog.com/entry/20060630/PSX_SEQ_Format) and a few threads on the PSXDevNetwork forum (this one specifically http://www.psxdev.net/forum/viewtopic.php?t=3500). If you know any more, no matter the language, then please let me know! I'd appreciate every bit of help!

(One VGMTrans unrelated question of mine would also be how similar MIDI and SEQ really is. I've seen people say that SEQ is exactly the same as MIDI, just with a different header. But the more I look into it the more it seems like SEQ was only based on MIDI but is kind of still a separate thing. Feel free to ignore this, though, if it's too off-topic!)

**Comment by sykhro** [comment]
At least about vgmtrans there's no documentation whatsoever besides the code, also we don't really do intermediate translation. Every format parser scans the file reading it command-by-command and instantiating the right SeqEvent depending on the data read. Of course in order to do this the parser has to know precisely how commands are structured and what they do (see PS1Seq.cpp) for an example. 
When a file is saved as midi the SeqEvents are read and their data is used to generate midi events. If you look at the `refactor` branch you should be able to identify the components more easily. I never dealt with PSX much so I can't really help

**Comment by loveemu** [comment]
Same with sykhro. Here is a part of my answer quoted from my Twitter DM.

There is no documentation explaining how VGMTrans interprets the bytes in a SEQ file, but I think the VGMTrans GUI itself can display the file details nicely. Double-click PS1 SEQ to display the hex viewer to see a list of highlighted events. Clicking on an individual event will also display some information on the toolbar.

For more details, you need to refer to the VGMTrans source code. However, for PS1 SEQ, the format is really almost the same as Standard MIDI File, so you don't need to read the source code much in my opinion.

**Comment by loveemu** [comment]
Another point, I can explain why the MIDI file size increases from the original SEQ file. Since VGMTrans is a general-purpose tool that is not specific to PS1 SEQ, it saves MIDI files that do not use running status for PS1 SEQ files that use running status.

If you want to get a MIDI file that is closer to the original PS1 SEQ file, export the SEQ file with a tool such as VGMTrans and then use another MIDI converter such as my seq2mid. <github.com/loveemu/seq2mid>

**Comment by loveemu** [comment]
> In the same vein, is there any documentation on how some custom PS1 audio drivers work (what drivers there are, which commands they accept, what those commands do etc.) or what the differences to generic SEQ are?

Unfortunately, I can't help with this. I can't give you much information other than my own materials.
